File: app/services/Markdown.js
import * as app from '@nativescript/core/application';

export default class Markdown {
  static markwon;
  static urlOpenCallback = (url) => console.log(url);
  static malformedHadingRegex = /^#+(?=[^#\s])/gm;

  static getInstance() {
    if (!this.markwon) {
      const activity = app.android.foregroundActivity;
      this.markwon = io.noties.markwon.Markwon.builder(activity)
          .usePlugin(this.getUrlPlugin())
          .usePlugin(this.getLinkResolverPlugin())
          .build();
    }
    return this.markwon;
  }

  static getUrlPlugin() {
    // https://developer.android.com/reference/android/text/util/Linkify
    const Pattern = java.util.regex.Pattern;
    // Match users (/u(ser)?/username) and subreddits (/r/subreddit) and valid urls
    const mask = Pattern.compile('(\\/(r|u|user)\\/[^\\s;.,:]+)|((?:^|[\\W])((ht|f)tp(s?):\\/\\/|www\\.)' +
      '(([\\w\\-]+\\.){1,}?([\\w\\-.~]+\\/?)*' +
      '[\\p{Alnum}.,%_=?&#\\-+()\\[\\]\\*$~@!:/{};\']*))',
    Pattern.CASE_INSENSITIVE | Pattern.MULTILINE | Pattern.DOTALL);
    const transformFilter = new android.text.util.Linkify.TransformFilter({
      transformUrl(match, url) {
        if (url.startsWith('/r/') || url.startsWith('/u/') || url.startsWith('/user/')) {
          return 'https://reddit.com' + url;
        }
        return url;
      },
    });
    const onClick = (url) => this.urlOpenCallback(url);
    const ClickableSpan = android.text.style.ClickableSpan.extend({
      url: null,
      setURL(url) {
        this.url = url;
      },
      onClick(view) {
        onClick(this.url);
      },
    });
    const LinkifyTextAddedListener = new io.noties.markwon.core.CorePlugin.OnTextAddedListener({
      builder: new android.text.SpannableStringBuilder(),
      onTextAdded(visitor, text, start) {
        this.builder.clear();
        this.builder.clearSpans();
        this.builder.append(text);
        if (android.text.util.Linkify.addLinks(this.builder, mask, null, null, transformFilter)) {
          const spans = this.builder.getSpans(0, this.builder.length(), java.lang.Object.class);
          if (spans != null && spans.length > 0) {
            const spannableBuilder = visitor.builder();
            for (let i = 0; i < spans.length; i++) {
              const span = spans[i];
              const clickableSpan = new ClickableSpan();
              clickableSpan.setURL(span.getURL());
              spannableBuilder.setSpan(
                  clickableSpan,
                  start + this.builder.getSpanStart(span),
                  start + this.builder.getSpanEnd(span),
                  this.builder.getSpanFlags(span));
            }
          }
        }
      },
    });
    const UrlResolverPlugin = io.noties.markwon.AbstractMarkwonPlugin.extend({
      configure(registry) {
        registry.require(io.noties.markwon.core.CorePlugin.class, new io.noties.markwon.MarkwonPlugin.Action({
          apply(corePlugin) {
            corePlugin.addOnTextAddedListener(LinkifyTextAddedListener);
          },
        }));
      },
    });
    return new UrlResolverPlugin();
  }

  static getLinkResolverPlugin() {
    const resolve = (view, link) => this.urlOpenCallback(link);
    const linkResolver = new io.noties.markwon.LinkResolver({resolve});
    const LinkResolverPlugin = io.noties.markwon.AbstractMarkwonPlugin.extend({
      configureConfiguration(builder) {
        builder.linkResolver(linkResolver);
      },
    });
    return new LinkResolverPlugin();
  }

  static setUrlOpenCallback(urlOpenCallback) {
    this.urlOpenCallback = urlOpenCallback;
  }

  static toMarkdown(text) {
    return this.getInstance().toMarkdown(this.fixMarkdownHeadings(text));
  }

  static setMarkdown(tv, text) {
    this.getInstance().setMarkdown(tv, this.fixMarkdownHeadings(text));
  }

  static setParsedMarkdown(tv, spanned) {
    this.getInstance().setParsedMarkdown(tv, spanned);
  }

  static fixMarkdownHeadings(text) {
    return (text || '').replace(this.malformedHadingRegex, (m) => m + ' ');
  }

  static setOnTouchListener(tv, listener) {
    tv.setOnTouchListener(new android.view.View.OnTouchListener({
      onTouch(v, event) {
        listener(event);
        return false;
      },
    }));
  }

  static setOnClickListener(tv, listener) {
    tv.setOnClickListener(new android.view.View.OnClickListener({
      onClick(v) {
        listener();
      },
    }));
  }

  static setSpannableFactory(tv) {
    tv.setSpannableFactory(io.noties.markwon.utils.NoCopySpannableFactory.getInstance());
  }
}

File: app/src/main/java/de/max/roehrl/vueddit2/service/Markdown.kt
package de.max.roehrl.vueddit2.service

import android.content.Context
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.ClickableSpan
import android.text.style.URLSpan
import android.text.util.Linkify
import android.view.View
import android.widget.TextView
import androidx.core.content.ContextCompat
import de.max.roehrl.vueddit2.R
import io.noties.markwon.*
import io.noties.markwon.core.CorePlugin
import io.noties.markwon.core.MarkwonTheme
import java.util.regex.Pattern

class Markdown private constructor(context: Context) {
    private val markwon: Markwon
    private val linkColor = ContextCompat.getColor(context, R.color.highlight)
    var urlOpenCallback = { url: String -> Url.openUrl(context, null, url) }
    private val malformedHeadingRegex = Regex("^#+(?=[^#\\s])", RegexOption.MULTILINE)

    init {
        markwon = Markwon.builder(context)
            .usePlugin(getUrlPlugin())
            .usePlugin(getLinkResolverPlugin())
            .build()
    }

    companion object : SingletonHolder<Markdown, Context>(::Markdown)

    private fun getUrlPlugin(): MarkwonPlugin {
        // https://developer.android.com/reference/android/text/util/Linkify
        // Match users (/u(ser)?/username) and subreddits (/r/subreddit) and valid urls
        val mask = Pattern.compile(
            "(\\/(r|u|user)\\/[^\\s;.,:]+)|((?:^|[\\W])((ht|f)tp(s?):\\/\\/|www\\.)" +
                    "(([\\w\\-]+\\.){1,}?([\\w\\-.~]+\\/?)*" +
                    "[\\p{Alnum}.,%_=?&#\\-+()\\[\\]\\*$~@!:/{};\']*))",
            Pattern.CASE_INSENSITIVE or Pattern.MULTILINE or Pattern.DOTALL
        )
        val transformFilter = Linkify.TransformFilter { _, url ->
            if (url.startsWith("/r/")
                || url.startsWith("/u/")
                || url.startsWith("/user/")
            ) {
                return@TransformFilter "${Reddit.api}$url"
            }
            return@TransformFilter url
        }

        class ClickableUrlSpan : ClickableSpan() {
            lateinit var url: String

            override fun onClick(view: View) {
                urlOpenCallback.invoke(url)
            }
        }

        class LinkifyTextAddedListener : CorePlugin.OnTextAddedListener {
            val builder = SpannableStringBuilder()
            override fun onTextAdded(visitor: MarkwonVisitor, text: String, start: Int) {
                builder.clear()
                builder.clearSpans()
                builder.append(text)
                if (Linkify.addLinks(builder, mask, null, null, transformFilter)) {
                    val spans: Array<URLSpan>? =
                        builder.getSpans(0, builder.length, URLSpan::class.java)
                    if (spans != null && spans.isNotEmpty()) {
                        val spannableBuilder = visitor.builder()
                        for (span in spans) {
                            val clickableSpan = ClickableUrlSpan()
                            clickableSpan.url = span.url
                            spannableBuilder.setSpan(
                                clickableSpan,
                                start + builder.getSpanStart(span),
                                start + builder.getSpanEnd(span),
                                builder.getSpanFlags(span)
                            )
                        }
                    }
                }
            }
        }

        class UrlResolverPlugin : AbstractMarkwonPlugin() {
            override fun configure(registry: MarkwonPlugin.Registry) {
                registry.require(CorePlugin::class.java) { corePlugin ->
                    corePlugin.addOnTextAddedListener(LinkifyTextAddedListener())
                }
            }

            override fun configureTheme(builder: MarkwonTheme.Builder) {
                builder.linkColor(linkColor)
            }
        }
        return UrlResolverPlugin()
    }

    private fun getLinkResolverPlugin(): MarkwonPlugin {
        class LinkResolverPlugin : AbstractMarkwonPlugin() {
            override fun configureConfiguration(builder: MarkwonConfiguration.Builder) {
                builder.linkResolver { _, url -> urlOpenCallback.invoke(url) }
            }
        }
        return LinkResolverPlugin()
    }

    fun setMarkdown(tv: TextView, text: String) {
        markwon.setMarkdown(tv, fixMarkdownSpacing(text))
    }

    fun toMarkDown(text: String): Spanned {
        return markwon.toMarkdown(fixMarkdownSpacing(text))
    }

    fun setMarkdown(tv: TextView, markdown: Spanned?) {
        markwon.setParsedMarkdown(tv, markdown!!)
    }

    private fun fixMarkdownSpacing(text: String): String {
        return text
            .replace(malformedHeadingRegex) { it.value + " " }
            .replace("]\n(", "](")
    }
}